pandas/_libs/parsers.pyx

```diff
@@ -1267,19 +1267,49 @@ cdef class TextReader:
             return self._string_convert(i, start, end, na_filter,
                                         na_hashset)
         elif is_categorical_dtype(dtype):
+            # TODO: I suspect that _categorical_convert could be
+            # optimized when dtype is an instance of CategoricalDtype
             codes, cats, na_count = _categorical_convert(
                 self.parser, i, start, end, na_filter,
                 na_hashset, self.c_encoding)
-            # sort categories and recode if necessary
             cats = Index(cats)
-            if not cats.is_monotonic_increasing:
+
+            # Determine if we should convert inferred string
+            # categories to a specialized type
+            if (isinstance(dtype, CategoricalDtype) and
+                    dtype.categories is not None):
+                if dtype.categories.is_numeric():
+                    # is ignore correct?
+                    cats = to_numeric(cats, errors='ignore')
+                elif dtype.categories.is_all_dates:
+                    # is ignore correct?
+                    if is_datetime64_dtype(dtype.categories):
+                        cats = to_datetime(cats, errors='ignore')
+                    else:
+                        cats = to_timedelta(cats, errors='ignore')
+
+            if (isinstance(dtype, CategoricalDtype) and
+                    dtype.categories is not None):
+                # recode for dtype.categories
+                categories = dtype.categories
```

Comment:
use [`_recode_for_categories`](https://github.com/pandas-dev/pandas/blob/66f4cc11725fd244c029058e9ac4743c4f36e2b4/pandas/core/categorical.py#L2148) here?

Reply:
Fixed (will wait to push until I hear back about https://github.com/pandas-dev/pandas/pull/17643#discussion_r141156526)